### PR TITLE
Deprecate v1alpha1

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
The purpose of this PR is deprecating v1alpha1 but since the current controller-tools does not support it, it also upgrades it to v0.6.1 the latest where it's available https://github.com/kubernetes-sigs/controller-tools/commit/2fc390d77e37de1019f0e8f2ab83017dfbc74f8a

### What issues does this PR fix or reference?
It's part of https://github.com/devfile/devworkspace-operator/issues/471, other part will be on DWO side

### Is your PR tested? Consider putting some instruction how to test your changes
It's tested with installing DWO and making sure everything works as expected. More see the corresponding section in https://github.com/devfile/devworkspace-operator/pull/478

### CODE REVIEW

File: schemas/latest/devfile.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "attributes": {
       "description": "Map of implementation-dependant free-form YAML attributes.",
       "type": "object",
-      "additionalProperties": true
+      "additionalProperties": {}
```
Comment (sleshchenko) [inline schemas/latest/devfile.json:12]: TODO: Take a look if it's a proper change or fix is needed in generator to revert it back

File: crds/workspace.devfile.io_devworkspaces.v1beta1.yaml [hunk 2]
```diff
@@ -4141,6 +4141,8 @@ spec:
                   of a devworkspace template.
                 properties:
                   attributes:
+                    additionalProperties:
```
Comment (sleshchenko) [inline crds/workspace.devfile.io_devworkspaces.v1beta1.yaml:4145]: TODO investigate `x-kubernetes-preserve-unknown-fields` as `additionalProperties.x-kubernetes-preserve-unknown-fields`. Do we need both?

File: generator/validate/gen.go [hunk 3]
```diff
@@ -34,11 +34,7 @@ func (Generator) RegisterMarkers(into *markers.Registry) error {
 // Generate validates the source code
 func (g Generator) Generate(ctx *genall.GenerationContext) error {
 	for _, root := range ctx.Roots {
-		ctx.Checker.Check(root, func(node ast.Node) bool {
-			// ignore interfaces
-			_, isIface := node.(*ast.InterfaceType)
-			return !isIface
-		})
+		//ctx.Checker.Check(root)
```
Comment (amisevsk) [inline generator/validate/gen.go:37]: Is this intentional?

File: schemas/latest/dev-workspace-template-spec.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
     "attributes": {
       "description": "Map of implementation-dependant free-form YAML attributes.",
       "type": "object",
-      "additionalProperties": true
+      "additionalProperties": {}
```
Comment (amisevsk) [inline schemas/latest/dev-workspace-template-spec.json:9]: This looks like an error

File: generator/validate/gen.go (hunk 3 above)
Comment (sleshchenko) [reply generator/validate/gen.go:37]: Good catch. I did it to fix nil pointer exception before I found the proper way to fix it. Reverting.

File: crds/workspace.devfile.io_devworkspaces.v1beta1.yaml (hunk 2 above)
Comment (sleshchenko) [reply crds/workspace.devfile.io_devworkspaces.v1beta1.yaml:4145]: Well, I tried to avoid having this but seems it's OK and related to the latest operator-sdk version, like for DWR we have it for a long time https://github.com/devfile/devworkspace-operator/pull/303

File: schemas/latest/dev-workspace-template-spec.json (hunk 4 above)
Comment (sleshchenko) [reply schemas/latest/dev-workspace-template-spec.json]: it's fixed in the first commit, see schemas/gen.go file. Sorry for squashing everything, otherwise it would be even difficult to review overall picture I believe.

So, AFAIU controller-gen changed their default so now they provide empty schema with XPreserveUnknownFields, which David (I don't know reason well, probably JSON schema does not have such field) ignores during generating JSON Schema.
but schema takes over allow in JSONSchemaPropsOrBool, so it led to empty object in additional properties.

Now, I ignore schema in additional properties if it does not define type, which is basic field.

**Review by amisevsk — APPROVED** [review]
LGTM. Would like @davidfestal's review as well, as he's more familiar with the generator code.

File: generator/crds/gen.go [hunk 5]
```diff
@@ -158,11 +165,30 @@ func (g Generator) Generate(ctx *genall.GenerationContext) error {
...
-			if err := ctx.WriteYAML(fileName, crd); err != nil {
+			if err := ctx.WriteYAML(fileName, extCrd); err != nil {
 				return err
 			}
 		}
 	}
 
 	return nil
 }
+
```
Comment (davidfestal) [inline generator/crds/gen.go:176]: Why is this necessary now ?

File: generator/schemas/gen.go [hunk 6]
```diff
@@ -238,16 +241,41 @@ This is not the case in the "%s' API group:
...
 			})
 
-			// Add the additionalProperies required to reflect the expected behavior from the K8S API,
+			// Add the additionalProperties required to reflect the expected behavior from the K8S API,
 			// (preserve-unknown-fields false by default unless on the Devfile metadata)
 			genutils.EditJSONSchema(&currentJSONSchema, func(schema *apiext.JSONSchemaProps) (newVisitor genutils.Visitor, stop bool) {
 				if schema == nil ||
-					schema.Type != "object" ||
-					schema.AdditionalProperties != nil {
+					schema.Type != "object" {
```
Comment (davidfestal) [inline generator/schemas/gen.go:248]: Not sure I understand the logic and requirement behind this change. Could you please provide some more context (I assume I'm missing some of the history here...)

File: generator/schemas/gen.go [hunk 7]
```diff
@@ -364,7 +395,13 @@ func addMarkdownDescription(orderedMap *orderedmap.OrderedMap) {
 	}
 	description, descriptionExists := orderedMap.Get("description")
 	if descriptionExists {
-		orderedMap.Set("markdownDescription", description)
+		s := fmt.Sprintf("%v", description)
+		if strings.HasPrefix(s, "Map of implementation-dependant") {
```
Comment (davidfestal) [inline generator/schemas/gen.go:399]: Am I missing something, or the same thing is done in both conditions  ?

File: schemas/latest/dev-workspace-template.json [hunk 8]
```diff
@@ -81,7 +81,8 @@
               },
               "fieldsV1": {
                 "description": "FieldsV1 holds the first JSON version format as described in the \"FieldsV1\" type.",
-                "type": "Any"
+                "type": "object",
+                "additionalProperties": false
```
Comment (davidfestal) [inline schemas/latest/dev-workspace-template.json:85]: Is it expected to have an `object` with no properties and additionalproperties to `false`? That seems strange because if I'm not mistaken it just prevents putting anything in the corresponding field. Maybe this should be managed as a special case (`maganedFields` is somehow a special case in itself...)?

File: generator/schemas/gen.go (hunk 7 above)
Comment (sleshchenko) [reply generator/schemas/gen.go:399]: :facepalm: it should be reverted, I crafted it for debugging. Good catch!

File: schemas/latest/dev-workspace-template.json (hunk 8 above)
Comment (sleshchenko) [reply schemas/latest/dev-workspace-template.json:85]: I'll take a look.

File: generator/crds/gen.go (hunk 5 above)
Comment (sleshchenko) [reply generator/crds/gen.go:176]: You mean filterTypesForCRDs but not a new line, right?

Comment (sleshchenko) [reply generator/crds/gen.go:176]: If `filterTypesForCRDs` probably we could filter nothing, but I thought it would be a good idea to do the same as controller-tools for CRDs does.

Comment (sleshchenko) [reply generator/crds/gen.go:176]: I can revert it and we'll add it only when it will be needed.

File: generator/schemas/gen.go (hunk 6 above)
Comment (sleshchenko) [reply generator/schemas/gen.go:248]: It's described in https://github.com/devfile/api/pull/524#discussion_r664532026
Duplicating

> So, AFAIU controller-gen changed their default so now they provide empty schema with XPreserveUnknownFields, which David (I don't know reason well, probably JSON schema does not have such field) ignores during generating JSON Schema.
but schema takes over allow in JSONSchemaPropsOrBool, so it led to empty object in additional properties.
> Now, I ignore schema in additional properties if it does not define type, which is basic field.

So, some debugging.
Before was: schema is empty:
Screenshot_20210708_123005
Now is, schema is set with just XPreserveUnknownFields:
Screenshot_20210708_122728

File: generator/schemas/gen.go (hunk 7 above)
Comment (sleshchenko) [reply generator/schemas/gen.go]: reverted

File: generator/crds/gen.go (hunk 5 above)
Comment (sleshchenko) [reply generator/crds/gen.go]: removed

File: schemas/latest/dev-workspace-template.json (hunk 8 above)
Comment (sleshchenko) [reply schemas/latest/dev-workspace-template.json]: Good catch!
It was tricky but I found a workaround https://github.com/devfile/api/pull/524/commits/20a45168c071c58fec1f049f1e1d4ae6629ecaa3

But I think additional work is needed out of the current PR scope since(the main branch state) we generate all metadata additionalProperties: false, like managedField https://github.com/devfile/api/blob/main/schemas/latest/dev-workspace-template.json#L100
But K8s Json Schema does not specify any and AFAIU it's supposed to be true by default https://kubernetesjsonschema.dev/v1.14.0/_definitions.json#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta
Screenshot_20210709_140215

**Comment by sleshchenko** [comment]
After discussion with @davidfestal I dropped adding metadata since seems I just misunderstood the referenced PR and it will lead to issues.

File: generator/schemas/gen.go (hunk 6 above)
Comment (davidfestal) [reply generator/schemas/gen.go:248]: Is this change still required now that the old generation behavior has been restored by PR https://github.com/devfile/api/pull/535 ?

Comment (sleshchenko) [reply generator/schemas/gen.go]: Thanks. After your PR it can be reverted.

**Review by davidfestal — APPROVED** [review]
Great !

Do you think you could open an issue about ensuring, in the `validate` generator, that all `Attributes` fields have the expected 3 annotations defined (cf PR https://github.com/devfile/api/pull/535) ?

**Comment by sleshchenko** [comment]
pushed with force where I just reorganize commits to have clear history
